<a href="https://colab.research.google.com/github/nug1209/PwC_Switzerland_Digital_Intelligence_Virtual_Case_Experience/blob/main/PwC_Task_3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In this task we deal with merchant loan data. We will create a forecast about how much will the total repayment will be after these loans are finished (after 30 months since the loans started). The formula for counting the repayment is provided. We will also create percentage values for each repayment in comparison with the origination amount.

Import libraries.

In [584]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from google.colab import drive
from datetime import datetime

Mount Google Drive to get the data.

In [585]:
drive.mount('/content/drive')
path = '/content/drive/MyDrive/Virtual Experience/PwC_Task_2_Data.csv'

df = pd.read_csv(path, delimiter=';')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Rename columns.

In [586]:
df = df.rename(columns={'Unnamed: 0':'origination_date', 'Origination Amount':'origination_amount'})
df.head(1)

origination_date  origination_amount  31.05.2019  30.06.2019  31.07.2019  \
0       31.05.2019         10018746.17  1443069.08  3332200.33  1328138.75   

   31.08.2019  30.09.2019  31.10.2019  30.11.2019  31.12.2019  ...  \
0   928085.74   736418.27   539403.31   427557.86   324459.32  ...   

   31.03.2020  30.04.2020  31.05.2020  30.06.2020  31.07.2020  31.08.2020  \
0   116684.68    92699.67    63399.66    53265.12    37121.13     29787.1   

   30.09.2020  31.10.2020  30.11.2020  31.12.2020  
0     24524.9    18085.94    16581.01    11442.97  

[1 rows x 22 columns]

In [587]:
df.shape

(20, 22)

Preparation for melting, i.e. transform wide format table to long format.

In [588]:
to_remove = ['origination_date', 'origination_amount']
value_columns = [i for i in list(df.columns) if i not in to_remove]

This was to calculate repayment percentage early in the processing, but may be it is better done later.

In [589]:
# for i in value_columns:
#   df[i + '_actual_repayment_percent'] = 100 * df[i] / df['origination_amount']

# df.columns

Convert to long format.

In [590]:
df = pd.melt(df, id_vars=['origination_date', 'origination_amount'], value_vars=value_columns)
df = df.rename(columns={'variable':'repayment_date', 'value':'repayment_amount'})

Remove dates with no repayment.

In [591]:
df = df[df['repayment_amount'] != 0.00]
# df

In [592]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 210 entries, 0 to 399
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   origination_date    210 non-null    object 
 1   origination_amount  210 non-null    float64
 2   repayment_date      210 non-null    object 
 3   repayment_amount    210 non-null    float64
dtypes: float64(2), object(2)
memory usage: 8.2+ KB


Add one row for the last entry as instructed in the task.

In [593]:
add_row = {'origination_date':['31.12.2020'], 'origination_amount':[30482978.52], 'repayment_date':['31.01.2021'], 'repayment_amount':[8747661.94]}

df = pd.concat([df, pd.DataFrame(add_row)], ignore_index=True)

In [594]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 211 entries, 0 to 210
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   origination_date    211 non-null    object 
 1   origination_amount  211 non-null    float64
 2   repayment_date      211 non-null    object 
 3   repayment_amount    211 non-null    float64
dtypes: float64(2), object(2)
memory usage: 6.7+ KB


Convert the date columns to date type.

In [595]:
df['origination_date'] = pd.to_datetime(df['origination_date'], format='%d.%m.%Y')
df['repayment_date'] = pd.to_datetime(df['repayment_date'], format='%d.%m.%Y')
# df.info()

Sort values.

In [596]:
df = df.sort_values(by=['origination_date', 'repayment_date'])
df

origination_date  origination_amount repayment_date  repayment_amount
0         2019-05-31         10018746.17     2019-05-31        1443069.08
1         2019-05-31         10018746.17     2019-06-30        3332200.33
3         2019-05-31         10018746.17     2019-07-31        1328138.75
6         2019-05-31         10018746.17     2019-08-31         928085.74
10        2019-05-31         10018746.17     2019-09-30         736418.27
..               ...                 ...            ...               ...
207       2020-10-31         27699586.46     2020-12-31        1503544.68
189       2020-11-30         29872889.68     2020-11-30        4383982.78
208       2020-11-30         29872889.68     2020-12-31        8383025.07
209       2020-12-31         30482978.52     2020-12-31        4373830.97
210       2020-12-31         30482978.52     2021-01-31        8747661.94

[211 rows x 4 columns]

In [597]:
df_copy = df.copy()

Create data frame from unique values of origination date and origination amount so that it can be merged later with the repayment values.

In [598]:
df1 = df[['origination_date', 'origination_amount']].drop_duplicates().reset_index(drop=True)
# df1

Get the first two repayments.

In [599]:
two_repayments = df.groupby('origination_date')['repayment_amount'].apply(list).reset_index(name='repayment')['repayment'].apply(lambda x: x[0: 2])
df2 = pd.DataFrame(two_repayments, columns=['repayment'])
# df2

In [600]:
df_agg = pd.concat([df1, df2], axis=1)
# df_agg

In [601]:
# def repayment_percent(a, b):
#   return [100* i / b for i in a]

def repayment_part(a, b):
  return [i / b for i in a]

In [602]:
df_agg['repayment_part'] = df_agg.apply(lambda row: repayment_part(row['repayment'], row['origination_amount']), axis=1)
df_agg

origination_date  origination_amount                 repayment  \
0        2019-05-31         10018746.17  [1443069.08, 3332200.33]   
1        2019-06-30         10868379.04   [1392751.6, 3011884.91]   
2        2019-07-31         10733932.61  [1537650.24, 2953335.55]   
3        2019-08-31         12558727.02   [1617681.94, 4082016.0]   
4        2019-09-30         14505071.44   [1992242.84, 3930445.6]   
5        2019-10-31         15652952.20  [2289453.76, 4682354.31]   
6        2019-11-30         15107713.30  [2162283.09, 4637701.69]   
7        2019-12-31         17004745.04  [2402403.37, 4947764.21]   
8        2020-01-31         16794379.95  [2502066.86, 4696910.48]   
9        2020-02-29         19217205.82  [2833811.35, 6142911.08]   
10       2020-03-31         21628095.29  [2843285.54, 6228477.78]   
11       2020-04-30         22334728.88  [3332800.29, 6476251.76]   
12       2020-05-31         22393345.39  [3575896.21, 7636995.58]   
13       2020-06-30         24671655.23  [3746372.58, 8983764.16]   
14       2020-07-31         25478232.23  [3470149.61, 8030091.09]   
15       2020-08-31         25878820.95   [3510328.95, 8374134.7]   
16       2020-09-30         25887161.41  [3976808.32, 7065477.24]   
17       2020-10-31         27699586.46  [3667678.21, 8752706.73]   
18       2020-11-30         29872889.68  [4383982.78, 8383025.07]   
19       2020-12-31         30482978.52  [4373830.97, 8747661.94]   

                                repayment_part  
0   [0.14403689398989936, 0.33259654186847215]  
1      [0.1281471316811932, 0.277123653758767]  
2    [0.14325134094539466, 0.2751401240630669]  
3     [0.128809387880142, 0.32503421672429983]  
4   [0.13734801984539555, 0.27097044066678516]  
5    [0.14626338410462916, 0.2991355400676429]  
6   [0.14312444557708145, 0.30697575456373005]  
7   [0.14127841166385405, 0.29096373973037826]  
8   [0.14898238979046083, 0.27967156239072705]  
9     [0.1474621948967605, 0.3196568292777956]  
10   [0.1314625953823417, 0.28798087378872467]  
11   [0.14922053936300123, 0.2899633031050252]  
12    [0.1596856632058583, 0.3410386187054653]  
13   [0.15184925960883736, 0.3641330132189919]  
14   [0.13620056441411907, 0.3151745779499083]  
15   [0.1356448563395621, 0.32359027160393106]  
16   [0.15362087240912367, 0.2729336418194798]  
17     [0.13240913236363167, 0.31598690986378]  
18  [0.14675455996930525, 0.28062317237466533]  
19    [0.1434843700437709, 0.2869687400875418]

In [603]:
discount_per_month = 2.5 / 12
remaining_part_after_discount_per_month = 100 - discount_per_month
remaining_part_after_discount_per_month

99.79166666666667

In [604]:
def expected_repayment_part(p):
  
  for i in np.arange(3, 31):
        
    calc1 = sum(p[:-1])
    
    calc2 = 1 - calc1
    
    calc3 = (i - 1) / 30
    
    calc4 = 1 - calc2
    
    calc5 = 1 + (calc4 * calc2)
    
    calc6 = p[1] * np.log(calc5)
    
    pi = max(calc6, 0)
    
    pi = pi * remaining_part_after_discount_per_month / 100

    p.append(pi)

  
  return p
	

In [605]:
p = [0.14403689398989936, 0.33259654186847215]
expected_repayment_part(p)

[0.14403689398989936,
 0.33259654186847215,
 0.03858781898375524,
 0.07391714902831535,
 0.07400063142832149,
 0.0719456679118513,
 0.06691912142639268,
 0.059053806453683466,
 0.048915805175408646,
 0.03750672726593,
 0.02613372267253328,
 0.01611823563685961,
 0.008418883822298589,
 0.00335593018489901,
 0.000611988668876962,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0,
 0.0]

In [606]:
sum(expected_repayment_part(p))

1.002118924517497

In [607]:
df_agg['expected_repayment_part'] = df_agg['repayment_part'].apply(expected_repayment_part)
df_agg['total_expected_repayment_part'] = df_agg['expected_repayment_part'].apply(sum)
df_agg['total_expected_repayment'] = df_agg['total_expected_repayment_part'] * df_agg['origination_amount']
df_agg['margin'] = df_agg['total_expected_repayment'] - df_agg['origination_amount']

In [608]:
df_agg.head()

origination_date  origination_amount                 repayment  \
0       2019-05-31         10018746.17  [1443069.08, 3332200.33]   
1       2019-06-30         10868379.04   [1392751.6, 3011884.91]   
2       2019-07-31         10733932.61  [1537650.24, 2953335.55]   
3       2019-08-31         12558727.02   [1617681.94, 4082016.0]   
4       2019-09-30         14505071.44   [1992242.84, 3930445.6]   

                                      repayment_part  \
0  [0.14403689398989936, 0.33259654186847215, 0.0...   
1  [0.1281471316811932, 0.277123653758767, 0.0292...   
2  [0.14325134094539466, 0.2751401240630669, 0.03...   
3  [0.128809387880142, 0.32503421672429983, 0.034...   
4  [0.13734801984539555, 0.27097044066678516, 0.0...   

                             expected_repayment_part  \
0  [0.14403689398989936, 0.33259654186847215, 0.0...   
1  [0.1281471316811932, 0.277123653758767, 0.0292...   
2  [0.14325134094539466, 0.2751401240630669, 0.03...   
3  [0.128809387880142, 0.32503421672429983, 0.034...   
4  [0.13734801984539555, 0.27097044066678516, 0.0...   

   total_expected_repayment_part  total_expected_repayment        margin  
0                       1.002119              1.003998e+07  21228.966894  
1                       1.000050              1.086892e+07    540.448804  
2                       1.000035              1.073430e+07    370.607450  
3                       1.001604              1.257887e+07  20147.521865  
4                       1.000017              1.450531e+07    241.418372

In [609]:
df_agg['total_expected_repayment'].sum()

399171091.11928964

According to the instructions, the result should be close to 84993122.67. So there seems to be something wrong. I'll check.

In [610]:
# pd.set_option('display.float_format', lambda x: '%.5f' % x)
# pd.reset_option('^display.', silent=True)

I found possible error. The instruction was to get number of cashflow forecast for months after 2020. Meanwhile my calculation was for cashflow for the whole timeline.

In [611]:
df_agg2 = df_agg[['origination_date', 'origination_amount', 'expected_repayment_part']]
df_agg2.head()

origination_date  origination_amount  \
0       2019-05-31         10018746.17   
1       2019-06-30         10868379.04   
2       2019-07-31         10733932.61   
3       2019-08-31         12558727.02   
4       2019-09-30         14505071.44   

                             expected_repayment_part  
0  [0.14403689398989936, 0.33259654186847215, 0.0...  
1  [0.1281471316811932, 0.277123653758767, 0.0292...  
2  [0.14325134094539466, 0.2751401240630669, 0.03...  
3  [0.128809387880142, 0.32503421672429983, 0.034...  
4  [0.13734801984539555, 0.27097044066678516, 0.0...

In [612]:
len(df_agg2['expected_repayment_part'][10])

30

Because the actualized repayment is already 20 months. So we will just clip the first 20 months of forecast. And then we sum the rest.

In [613]:
# two_repayments = df.groupby('origination_date')['repayment_amount'].apply(list).reset_index(name='repayment')['repayment'].apply(lambda x: x[0: 2])
df_agg2['expected_repayment_part_to_forecast'] = df_agg2['expected_repayment_part'].apply(lambda x: list(x[20:]))

<ipython-input-613-4ba1820f8c4c>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_agg2['expected_repayment_part_to_forecast'] = df_agg2['expected_repayment_part'].apply(lambda x: list(x[20:]))


In [615]:
len(df_agg2['expected_repayment_part_to_forecast'][2])

10

In [617]:
df_agg2 = df_agg2.drop(columns=['expected_repayment_part'])
df_agg2.head()

origination_date  origination_amount  \
0       2019-05-31         10018746.17   
1       2019-06-30         10868379.04   
2       2019-07-31         10733932.61   
3       2019-08-31         12558727.02   
4       2019-09-30         14505071.44   

                 expected_repayment_part_to_forecast  
0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  
1  [0.00035375557378911494, 0.0001274502637119302...  
2  [0.0003363070416652008, 0.000127581610524241, ...  
3  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...  
4  [0.0005597463310640909, 0.00025435046008442735...

In [618]:
df_agg2['total_expected_repayment_part_to_forecast'] = df_agg2['expected_repayment_part_to_forecast'].apply(sum)
df_agg2['total_expected_repayment_to_forecast'] = df_agg2['total_expected_repayment_part_to_forecast'] * df_agg2['origination_amount']
df_agg2.head()


origination_date  origination_amount  \
0       2019-05-31         10018746.17   
1       2019-06-30         10868379.04   
2       2019-07-31         10733932.61   
3       2019-08-31         12558727.02   
4       2019-09-30         14505071.44   

                 expected_repayment_part_to_forecast  \
0  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   
1  [0.00035375557378911494, 0.0001274502637119302...   
2  [0.0003363070416652008, 0.000127581610524241, ...   
3  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   
4  [0.0005597463310640909, 0.00025435046008442735...   

   total_expected_repayment_part_to_forecast  \
0                                   0.000000   
1                                   0.000511   
2                                   0.000500   
3                                   0.000000   
4                                   0.000959   

   total_expected_repayment_to_forecast  
0                              0.000000  
1                           5552.760346  
2                           5361.769074  
3                              0.000000  
4                          13904.534339

In [619]:
df_agg2['total_expected_repayment_to_forecast'].sum()

48330.86427843139

The target number from the instruction is 84993122.67. I got 48330.86. Still wrong. Sorry but I'm going to let this go. I'm moving on to machine learning problems that are the objectives of my attempts in doing these virtual experiences. I'm making machine learning portfolio.